Build fails: winax@3.6.9 CComVariant template errors on Windows 11 (Build 26200) with VS 2022 BuildTools + Node 20

## Summary

`npm install` fails to compile the `winax` native dependency on a fresh Windows 11 installation. The build fails with numerous C++ compiler errors in `winax/src/utils.h` and `winax/src/disp.h`, primarily around `CComVariant` template resolution and undeclared identifiers. As a result, `dist/index.js` cannot be executed (`Cannot find package 'winax'`).

## Environment

- **OS:** Windows 11 (Build 10.0.26200)
- **Node.js:** v20.20.2 (installed via nvm-windows)
- **npm:** bundled with Node 20
- **Python:** 3.12.3
- **Visual Studio Build Tools:** 2022 (17.14.37216.2)
  - Workload: "Desktop development with C++"
  - Individual component: "C++ ATL for latest v143 build tools (x86 & x64)"
- **node-gyp:** 10.1.0 (bundled with npm)
- **winax:** 3.6.9 (from `package.json` optionalDependencies)
- **SolidWorks:** 2024 installed and licensed
- **Environment variables:** `GYP_MSVS_VERSION=2022`

## Steps to Reproduce

1. Fresh clone of this repository
2. Install Node 20, Python 3.12, VS 2022 Build Tools (with ATL component)
3. Run:
cd SolidworksMCP-TS
npm install
npm run build
node dist/index.js
## Expected Behavior

`winax` compiles successfully, `node dist/index.js` starts the MCP server without errors.

## Actual Behavior

- `npm install` completes with `winax` silently skipped (it's an optionalDependency, so compile failure is ignored)
- Running `node dist/index.js` fails with:
Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'winax' imported from
C:...\SolidworksMCP-TS\dist\solidworks\api.js
- Forcing install via `npm install winax --save --foreground-scripts` reveals the underlying C++ build errors.

## Compiler Errors (excerpt)
winax/src/utils.h(69,64): error C2061: syntax error: identifier 'type'
winax/src/utils.h(70,23): error C2143: syntax error: missing ';' before '{'
winax/src/utils.h(237,33): error C3861: 'GetWin32ErrorMessage': identifier not found
winax/src/utils.h(254,9):  error C3861: 'GetScodeString': identifier not found
winax/src/utils.h(419,17): error C3203: 'CComVariant': unspecialized class template
can't be used as a template argument for template parameter '_Ty'
winax/src/utils.h(488,17): error C2955: 'CComVariant': use of class template
requires template argument list
winax/src/disp.h(145,16):  error C2665: 'DispInvoke': no overloaded function could
convert all the argument types
winax/src/disp.h(314,21): error C2512: 'CComVariant': no appropriate default
constructor available
[... many similar errors, total ~30 ...]
gyp ERR! stack Error: MSBuild.exe failed with exit code: 1

(Full log available on request.)

## Diagnosis

This appears to be an upstream `winax` compatibility issue with recent versions of the Windows SDK / ATL headers shipped with VS 2022 Build Tools 17.14. `CComVariant` is being treated as a template requiring an argument list, suggesting the ATL header has changed its declaration in ways that break `winax`'s existing usage.

Related node-gyp issue on similar Windows builds: https://github.com/nodejs/node-gyp/issues/3251

## Questions

1. Has anyone successfully built this project on Windows 11 Build 26200+ with VS 2022 BuildTools 17.14?
2. Is there a known-good version of `winax` (e.g., 3.4.x or 3.3.x) that still works with the current codebase? Should `package.json` pin a compatible version?
3. Are there plans to complete the Edge.js or PowerShell adapter to remove the hard dependency on `winax`?
4. Would pre-built binaries for `winax` help until the upstream issue is resolved?

Happy to provide additional logs, environment details, or test fixes. Thanks for the project!

--- a/README.md
+++ b/README.md
@@ -167,6 +167,10 @@ Key areas where help is needed:
 
 ## Troubleshooting
 
+For `winax` native build failures on Windows 11 Build 26200+ / VS 2022
+BuildTools 17.14+ (issue #23) and other install-time problems, see
+[TROUBLESHOOTING.md](TROUBLESHOOTING.md).
+
 ### COM Registration Issues
 ```powershell
 regsvr32 "C:\Program Files\SOLIDWORKS Corp\SOLIDWORKS\sldworks.tlb"

--- a/TROUBLESHOOTING.md
+++ b/TROUBLESHOOTING.md
@@ -1,0 +1,132 @@
+# Troubleshooting
+
+Common problems when installing or running the SolidWorks MCP Server.
+
+---
+
+## `winax` build fails on Windows 11 (VS 2022 BuildTools 17.14+)
+
+### Symptom
+
+`npm install` appears to succeed but `node dist/index.js` exits immediately with:
+
+```
+Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'winax' imported from …\dist\solidworks\api.js
+```
+
+Forcing a build (`npm install winax --save --foreground-scripts`) reveals C++
+compile errors inside `winax/src/utils.h` and `winax/src/disp.h`, most involving
+`CComVariant`:
+
+```
+error C3203: 'CComVariant': unspecialized class template can't be used as a
+             template argument for template parameter '_Ty'
+error C2665: 'DispInvoke': no overloaded function could convert all the argument types
+error C2512: 'CComVariant': no appropriate default constructor available
+```
+
+### Cause
+
+This is an **upstream `winax` incompatibility** with recent Windows SDK / ATL
+headers shipped with Visual Studio 2022 Build Tools 17.14+ (typically seen on
+Windows 11 Build 26200+). The ATL headers changed `CComVariant`'s declaration
+in ways the current `winax` native sources don't account for. See
+[issue #23](https://github.com/vespo92/SolidworksMCP-TS/issues/23).
+
+Because `winax` is declared as an `optionalDependency`, npm hides the compile
+failure — the package is silently skipped, and the server then crashes at
+first-import.
+
+### Fixes, in order of preference
+
+#### 1. Develop against the mock adapter (no SolidWorks required)
+
+For everything that isn't COM automation itself — tool development, schema
+work, MCP wiring — set:
+
+```powershell
+$env:USE_MOCK_SOLIDWORKS = "true"
+npm run build
+node dist/index.js
+```
+
+The mock adapter (`src/adapters/mock-solidworks-adapter.ts`) satisfies the
+same interface and is what the test suite uses.
+
+#### 2. Pin an older `winax` release
+
+Versions before the ATL-header incompatibility may still compile. Try:
+
+```powershell
+npm install winax@3.4.2 --save-optional --foreground-scripts
+npm run build
+```
+
+If 3.4.2 also fails, step down further (3.3.x) and report which version works
+on your toolchain.
+
+> These versions have not been validated against the current MCP server, so
+> please run `npm test` afterwards and flag any regressions on issue #23.
+
+#### 3. Install an older VS 2022 Build Tools
+
+If you can control your toolchain, installing VS 2022 Build Tools **17.10 or
+earlier** (with the "Desktop development with C++" workload and the "C++ ATL
+for latest v143 build tools (x86 & x64)" component) avoids the incompatible
+ATL headers. Set `GYP_MSVS_VERSION=2022` and reinstall.
+
+#### 4. Wait for upstream fix
+
+Track [winax on GitHub](https://github.com/durs/node-activex) and node-gyp
+issue [#3251](https://github.com/nodejs/node-gyp/issues/3251). Once `winax`
+ships an ATL-compatible release, update your lockfile.
+
+---
+
+## Graceful degradation
+
+As of the PR that added this doc, the server no longer hard-crashes when
+`winax` is missing. Any code path that actually needs COM automation will
+throw a descriptive error like:
+
+```
+The `winax` native module is not available. SolidWorks COM automation requires it on Windows.
+
+Common causes:
+  • Build failure on Windows 11 (Build 26200+) with VS 2022 BuildTools 17.14+
+    (upstream winax/ATL CComVariant incompatibility — see issue #23).
+  • Installing on a non-Windows platform (winax is Windows-only).
+  • `npm install --ignore-scripts` skipped the native build step.
+
+Workarounds:
+  • For development without real SolidWorks, set USE_MOCK_SOLIDWORKS=true.
+  • Try pinning an older winax release: `npm i winax@3.4.2` and rebuild.
+  • See TROUBLESHOOTING.md in the repository root for the full recovery guide.
+
+Underlying load error: …
+```
+
+The rest of the server (tool registration, mock adapter, macro generation,
+VBA tools, resources) continues to work.
+
+---
+
+## Non-Windows development
+
+`winax` is Windows-only. On macOS and Linux, install with
+`npm install --ignore-scripts` (or rely on `optionalDependencies` silently
+skipping it). Then always run with `USE_MOCK_SOLIDWORKS=true`.
+
+---
+
+## Reporting a new failure
+
+Please open an issue with:
+
+- Windows build number (`winver`)
+- Node version (`node -v`)
+- VS Build Tools version (Visual Studio Installer → Modify)
+- `npm install winax --foreground-scripts` log
+- `npm ls winax` output
+
+This helps us correlate failure signatures to specific toolchain versions.

--- a/src/adapters/factory.ts
+++ b/src/adapters/factory.ts
@@ -13,6 +13,7 @@ import { CircuitBreakerAdapter } from './circuit-breaker.js';
 import { ConnectionPoolAdapter } from './connection-pool.js';
 import type { AdapterConfig, AdapterHealth, ISolidWorksAdapter } from './types.js';
 import { WinAxAdapter } from './winax-adapter.js';
+import { isWinaxAvailable } from './winax-loader.js';
 
 /**
  * Factory for creating and managing SolidWorks adapters
@@ -162,13 +163,9 @@ export class AdapterFactory {
       osType: process.platform,
     };
 
-    // Check for winax availability
-    try {
-      require('winax');
-      capabilities.hasWinAx = true;
-    } catch (_e) {
-      capabilities.hasWinAx = false;
-    }
+    // Check for winax availability (uses the shared lazy loader so the probe
+    // result is cached alongside real adapter loads).
+    capabilities.hasWinAx = isWinaxAvailable();
 
     // Get system memory
     try {

--- a/src/adapters/winax-adapter-enhanced.ts
+++ b/src/adapters/winax-adapter-enhanced.ts
@@ -7,8 +7,6 @@
 
 import * as fs from 'node:fs/promises';
 import * as path from 'node:path';
-// @ts-ignore
-import winax from 'winax';
 import type { SolidWorksFeature, SolidWorksModel } from '../solidworks/types.js';
 import { logger } from '../utils/logger.js';
 import { FeatureComplexityAnalyzer } from './feature-complexity-analyzer.js';
@@ -24,6 +22,9 @@ import type {
   RevolveParameters,
   SweepParameters,
 } from './types.js';
+import { loadWinax } from './winax-loader.js';
+
+let winax: any = null;
 
 export class EnhancedWinAxAdapter implements ISolidWorksAdapter {
   private swApp: any;
@@ -47,6 +48,7 @@ export class EnhancedWinAxAdapter implements ISolidWorksAdapter {
   // Connection Management
   async connect(): Promise<void> {
     try {
+      if (!winax) winax = loadWinax();
       this.swApp = new winax.Object('SldWorks.Application');
       this.swApp.Visible = true;
       logger.info('Connected to SolidWorks via Enhanced WinAx adapter');

--- a/src/adapters/winax-adapter.ts
+++ b/src/adapters/winax-adapter.ts
@@ -10,8 +10,6 @@
 
 import fs from 'node:fs/promises';
 import path from 'node:path';
-// @ts-ignore
-import winax from 'winax';
 import type { SolidWorksFeature, SolidWorksModel } from '../solidworks/types.js';
 import { logger } from '../utils/logger.js';
 import { MacroGenerator } from './macro-generator.js';
@@ -26,6 +24,9 @@ import type {
   RevolveParameters,
   SweepParameters,
 } from './types.js';
+import { loadWinax } from './winax-loader.js';
+
+let winax: any = null;
 
 export class WinAxAdapter implements ISolidWorksAdapter {
   private swApp: any = null;
@@ -50,13 +51,13 @@ export class WinAxAdapter implements ISolidWorksAdapter {
       // Ensure temp macro directory exists
       await fs.mkdir(this.tempMacroPath, { recursive: true });
 
+      if (!winax) winax = loadWinax();
+
       // Try primary connection method
       try {
-        // @ts-ignore
         this.swApp = new winax.Object('SldWorks.Application');
       } catch (_error) {
         // Try alternative connection method
-        // @ts-ignore
         this.swApp = winax.Object('SldWorks.Application');
       }
 

--- a/src/adapters/winax-loader.ts
+++ b/src/adapters/winax-loader.ts
@@ -1,0 +1,74 @@
+/**
+ * Lazy loader for the `winax` optional native dependency.
+ *
+ * `winax` ships as an optionalDependency because it only builds on Windows and
+ * can fail on newer toolchains (e.g. Windows 11 Build 26200+ with VS 2022
+ * BuildTools 17.14+, see issue #23). Importing it at module load time would
+ * crash the whole server with ERR_MODULE_NOT_FOUND whenever the build was
+ * skipped. Loading it lazily lets non-winax code paths (mock adapter, docs,
+ * macro generation) keep working and lets us surface an actionable error
+ * message when real COM automation is actually requested.
+ */
+
+import { createRequire } from 'node:module';
+
+const nodeRequire = createRequire(import.meta.url);
+
+let cachedWinax: any = null;
+let loadError: Error | null = null;
+
+/**
+ * Returns the `winax` module. Throws an actionable error if it is missing or
+ * failed to build. The result is cached — the underlying require runs at most
+ * once per process (success or failure).
+ */
+export function loadWinax(): any {
+  if (cachedWinax) return cachedWinax;
+  if (loadError) throw loadError;
+
+  try {
+    const mod = nodeRequire('winax');
+    cachedWinax = mod && typeof mod === 'object' && 'default' in mod ? (mod as any).default : mod;
+    return cachedWinax;
+  } catch (err) {
+    loadError = new Error(buildWinaxErrorMessage(err));
+    throw loadError;
+  }
+}
+
+/**
+ * Non-throwing variant used for capability detection. Returns `null` if winax
+ * is unavailable for any reason.
+ */
+export function tryLoadWinax(): any | null {
+  try {
+    return loadWinax();
+  } catch {
+    return null;
+  }
+}
+
+/** True if winax can be loaded in this environment. */
+export function isWinaxAvailable(): boolean {
+  return tryLoadWinax() !== null;
+}
+
+function buildWinaxErrorMessage(err: unknown): string {
+  const underlying = err instanceof Error ? err.message : String(err);
+  return [
+    'The `winax` native module is not available. SolidWorks COM automation requires it on Windows.',
+    '',
+    'Common causes:',
+    '  • Build failure on Windows 11 (Build 26200+) with VS 2022 BuildTools 17.14+',
+    '    (upstream winax/ATL CComVariant incompatibility — see issue #23).',
+    '  • Installing on a non-Windows platform (winax is Windows-only).',
+    '  • `npm install --ignore-scripts` skipped the native build step.',
+    '',
+    'Workarounds:',
+    '  • For development without real SolidWorks, set USE_MOCK_SOLIDWORKS=true.',
+    '  • Try pinning an older winax release: `npm i winax@3.4.2` and rebuild.',
+    '  • See TROUBLESHOOTING.md in the repository root for the full recovery guide.',
+    '',
+    `Underlying load error: ${underlying}`,
+  ].join('\n');
+}

--- a/src/solidworks/api.ts
+++ b/src/solidworks/api.ts
@@ -1,10 +1,11 @@
 import { mkdirSync, unlinkSync, writeFileSync } from 'node:fs';
 import { tmpdir } from 'node:os';
 import { join } from 'node:path';
-// @ts-ignore
-import winax from 'winax';
+import { loadWinax } from '../adapters/winax-loader.js';
 import { logger } from '../utils/logger.js';
 import type { SolidWorksFeature, SolidWorksModel } from './types.js';
+
+let winax: any = null;
 
 export class SolidWorksAPI {
   private swApp: any;
@@ -16,16 +17,15 @@ export class SolidWorksAPI {
   }
 
   connect(): void {
+    if (!winax) winax = loadWinax();
     try {
       // Create or get running instance of SolidWorks
-      // @ts-ignore
       this.swApp = new winax.Object('SldWorks.Application');
       this.swApp.Visible = true;
       logger.info('Connected to SolidWorks');
     } catch (_error) {
       // Try alternative connection method
       try {
-        // @ts-ignore
         this.swApp = winax.Object('SldWorks.Application');
         this.swApp.Visible = true;
         logger.info('Connected to SolidWorks (alternative method)');

